main/src/ca/uwaterloo/flix/language/phase/Deriver.scala

```diff
@@ -901,7 +901,7 @@ object Deriver {
     * Builds a string concatenation expression from the given expressions.
     */
   private def concat(exp1: KindedAst.Expr, exp2: KindedAst.Expr, loc: SourceLocation)(implicit flix: Flix): KindedAst.Expr = {
-    KindedAst.Expr.Binary(SemanticOp.StringOp.Concat, exp1, exp2, Type.freshVar(Kind.Star, loc), loc)
+    KindedAst.Expr.InvokeMethod2(exp1, Name.Ident("concat", loc), List(exp2), Type.freshVar(Kind.JvmConstructorOrMethod, loc), Type.freshVar(Kind.Star, loc), Type.freshVar(Kind.Eff, loc), loc)
```

Comment:
@mlutze The deriver is post type inference right, so we should have the specific types and effects here, right?

Reply:
Wrong. Deriver is pre-inference.

Reply:
This is the line causing the tests to fail?

Reply:
@mlutze Any idea why the tests are failing them? I thought it was because of specifying the wrong types.

Reply:
Just for reference, I was looking at this KindedAST.Expr method     `case class InvokeMethod2(exp: Expr, methodName: Name.Ident, exps: List[Expr], mvar: Type.Var, tvar: Type.Var, evar: Type.Var, loc: SourceLocation) extends Expr
`

Reply:
I believe the tests are failing due to the impurity of Java operations. To solve this, we need the expression to be `unsafe e1.concat(e2)`. The `unsafe` expression is internally called `MaskingCast`, I think, so we just need to wrap that around our concat expression here.

Reply:
Sounds good.

Reply:
I wrapped the concat expr with `KindedAST.Expr.UncheckedMaskingCast`